### PR TITLE
Python API for Complex Storage and storage copy logic

### PR DESCRIPTION
**Pull request by anjali411**
Following up on this: https://github.com/pytorch/pytorch/pull/35851 cross dtype storage copy is not being used internally, so I have not included cross dtype copy for complex.

### CODE REVIEW

**Review by izdeby — CHANGES_REQUESTED** [review]
Please fix tests and make sure that phabricator build passes as well.

File: aten/src/TH/THGenerateComplexFloatType.h [hunk 1]
```diff
@@ -0,0 +1,18 @@
+#ifndef TH_GENERIC_FILE
+#error "You must define TH_GENERIC_FILE before including THGenerateComplexFloatType.h"
+#endif
+
+#define scalar_t std::complex<float>
+#define accreal double
```
[the remark this answers is not in the document]
Comment (anjali411) [reply aten/src/TH/THGenerateComplexFloatType.h:6]: double

File: torch/csrc/utils/byte_order.cpp [hunk 2]
```diff
@@ -189,6 +190,40 @@ void THP_decodeDoubleBuffer(double* dst, const uint8_t* src, THPByteOrder order,
   }
 }
 
+void THP_decodeComplexFloatBuffer(std::complex<float>* dst, const uint8_t* src, THPByteOrder order, size_t len)
```
[the remark this answers is not in the document]
Comment (anjali411) [reply torch/csrc/utils/byte_order.cpp:193]: I am converting the complex tensors to float tensors so the decoding logic used for complex tensors is the same as float tensors

File: c10/util/TypeCast.h [hunk 3]
```diff
@@ -66,6 +66,16 @@ template <typename dest_value_t, typename src_value_t>
...
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+template <typename dest_value_t, typename src_value_t>
+  struct static_cast_with_inter_type<c10::complex<dest_value_t>, c10::complex<src_value_t>> {
+    C10_HOST_DEVICE static inline c10::complex<dest_value_t> apply(c10::complex<src_value_t> src) {
+      return c10::complex<dest_value_t>(src.real(), src.imag());
+    }
+};
+#endif
+
```
Comment (zasdfgbnm) [inline c10/util/TypeCast.h:78]: Is this necessary?

File: aten/src/THC/THCTensorCopy.cu [hunk 4]
```diff
@@ -10,9 +10,9 @@ template <typename T>
 struct CopyOp {
   __device__ __forceinline__ void operator()(T* dst, T* src) {
 #if __CUDA_ARCH__ >= 350
-    *dst = ScalarConvert<T, T>::to(__ldg(src));
+    *dst = c10::static_cast_with_inter_type<T, T>::apply(*src);
 #else
-    *dst = ScalarConvert<T, T>::to(*src);
+    *dst = c10::static_cast_with_inter_type<T, T>::apply(*src);
```
Comment (anjali411) [inline aten/src/THC/THCTensorCopy.cu:15]: cc. @zasdfgbnm

File: c10/util/TypeCast.h (hunk 3 above)
Comment (anjali411) [reply c10/util/TypeCast.h:78]: linked you to where I am using it

Comment (anjali411) [reply c10/util/TypeCast.h]: updated!

File: aten/src/THC/THCTensorCopy.cu (hunk 4 above)
Comment (izdeby) [inline aten/src/THC/THCTensorCopy.cu:15]: why is this needed?

File: errors.txt [hunk 5]
```diff
@@ -0,0 +1,971 @@
...
+-- Trying to find preferred BLAS backend of choice: MKL
+-- MKL_THREADING = OMP
+-- MKL libraries: /data/users/chourdiaanjali/miniconda3/envs/myenv2/lib/libmkl_intel_lp64.so;/data/users/chourdiaanjali/miniconda3/envs/myenv2/lib/libmkl_gnu_thread.so;/data/users/chourdiaanjali/miniconda3/envs/myenv2/lib/libmkl_core.so;-fopenmp;/usr/lib64/libpthread.so;/usr/lib64/libm.so;/usr/lib64/libdl.so
+-- MKL include directory: /data/users/chourdiaanjali/miniconda3/envs/myenv2/include
+-- MKL OpenMP type: GNU
+-- MKL OpenMP library: -fopenmp
+-- Found PythonInterp: /data/users/chourdiaanjali/miniconda3/envs/myenv2/bin/python (found version "3.6.9") 
+-- Failed to find LLVM FileCheck
+-- Using third party subdirectory Eigen.
+-- Found PythonInterp: /data/users/chourdiaanjali/miniconda3/envs/myenv2/bin/python (found suitable version "3.6.9", minimum required is "3.0")
```
Comment (izdeby) [inline errors.txt:20]: this should not be included

**Review by izdeby — APPROVED** [review]
Look good overall. just remove the log file

File: errors.txt (hunk 5 above)
Comment (anjali411) [reply errors.txt]: oops

File: aten/src/THC/THCTensorRandom.cu [hunk 6]
```diff
@@ -15,7 +15,7 @@
 // NB: ROCm compiler seems to have a bug where __host__ functions must be
 // explicitly specified extern "C" otherwise ROCm compiler doesn't respect it.
 // See https://github.com/RadeonOpenCompute/hcc/issues/839
-extern "C" __host__ void THCRandom_getRNGState(at::Generator gen_, THByteTensor *rng_state)
+__host__ void THCRandom_getRNGState(at::Generator gen_, THByteTensor *rng_state)
```
Comment (zasdfgbnm) [inline aten/src/THC/THCTensorRandom.cu:18]: You might also need to change line 43

Comment (anjali411) [reply aten/src/THC/THCTensorRandom.cu:18]: ah yeah just realized! thanks

**Pull request closed without merging**